rust/osm2lanes/src/transform/tags_to_lanes/road.rs

```diff
@@ -0,0 +1,412 @@
+use std::str::FromStr;
+
+use Highway::*;
+use Hunch::*;
+
+use crate::road::LaneDirection;
+use crate::tag::TagKey;
+use crate::transform::RoadMsg::Unimplemented;
+use crate::transform::{RoadFromTags, RoadMsg, RoadWarnings, Tags};
+use crate::Locale;
+
+pub fn tags_to_lanes(tags: &Tags, locale: &Locale) -> Result<RoadFromTags, RoadError> {
+    let mut road: Road = Road::new(locale);
+
+    road.set_tags::<Oneway>(tags);
+    road.set_tags::<Busway>(&tags);
+    road.set_tags::<Lanes>(&tags);
+
+    // if (guessing_mode)
+    {
+        road.guess::<Oneway>();
+        road.guess::<Busway>();
+        road.guess::<Lanes>();
+    }
+
+    road.calculate_lanes_ltr().into()
+}
+
+/// A collection of schemas. We have a Hunch about each schema because we need at least an option,
+/// so why not track it as a Hunch?
+struct Road {
+    busway: Hunch<Busway>,
+    lanes: Hunch<Lanes>,
+    // map_of_schema_to_hunches
+}
+
+impl Road {
+    /// Sets the Hunch for the schema to a Tagged* one, with tags applied.
+    fn set_tags<S: Scheme>(&mut self, tags: &Tags) {
+        let x = self.get_hunch::<S>();
+        self.set_hunch::<S>(x.tagged_to_be(x.ok().set_tags(&tags)))
+    }
+    fn get<T: Scheme>(&self) -> T {}
+    fn get_hunch<T: Scheme>(&self) -> Hunch<T> {} // Not Unknown
+    fn set_hunch<T: Scheme>(&self, scheme: Hunch<T>) {}
+
+    // pub fn busway(&self) -> &mut Busway {
+    //     match self.busway.some() {
+    //         None => {
+    //             self.busway = Assume(Busway::default(self));
+    //             self.busway.to_mut_ref();
+    //             &mut b
+    //         }
+    //         Some(b) => &mut b,
+    //     }
+    // }
+}
+
+type Problem = String;
+type Count = usize;
+type Qty = f64;
+enum Highway {
+    // Piste, // Track, // Bridleway,
+    Footway,
+    Cycleway,
+    Service,
+    Road,
+    Street,
+    Highway,
+    Freeway,
+}
+
+/// A value of different certainties. Hunches combine together into another hunch in predefined ways.
+enum Hunch<T> {
+    // Unknown,
+    Tagged(T),
+    Implied(T),
+    Assumed(T),
+    Guessed(T),
+}
```

Comment:
I think a question you had was asking how to proceed in integrating this. I would start by adding and using this new type in a limited scope and then slowly refactoring out.

Trying to do it all at once will be difficult to make it pan out.

Reply:
I agree, that's why I was happy to split any error tracking out of `Hunch`, to see what I wanted, which is `downgrade` and `reconcile`.

And thanks for being a persistent voice of reason :)